### DISCUSSION TITLE
`interpolant` function with opti variable input.

### DISCUSSION THREAD

**Discussion by mrrezaie** [casadi/casadi#3584] [Q&A]
Hi, I'm trying to use non-casadi functions to write my constraints, and I already know that `Callback` should be used, but I cannot handle the derivatives as I posted this earlier: https://github.com/casadi/casadi/discussions/3580

Indeed, the non-casadi functions create non-linear curves and my formulation with somehow equivalent exponential and polynomial functions works very well (converged very efficiently), but not good enough and I want to use the exact curves generated by the non-casadi functions. I was wondering if it's possible to do this with `interpolant` instead:

```python
x1 = np.linspace(0, 2, 101).tolist()
y1 = [non_casadi_function_1(i) for i in x1]
F1 = casadi.interpolant('F1', 'linear', [x1], y1)
print(F1(1.1)) # it works and prints a correct value, DM(0.961348)

x2 = np.linspace(-1, 1, 101).tolist()
y2 = [non_casadi_function_2(i) for i in x2]
F2 = casadi.interpolant('F2', 'linear', [x2], y2)
```
```python
opti = casadi.Opti()
A = opti.variable(n)
B = opti.variable(n)
print( F1(A[0]) ) # prints MX(F1(opti1_x_1[0]){0})

for i in range(n):
    opti.subject_to( F1(A[i]) == F2(B[i]) )
```

Does this example make sense? I don't get any error, but my objective is being maximized (never converged) which means something is wrong with my formulation.

Any help is greatly appreciated.

**Comment by mrrezaie** [comment]
Any comment? I cannot really understand what is wrong with this example.

**Comment by cdhainaut** [comment]
Can you detail the problem you have ?
As we don't have your "non_casadi_function_{i}" we can't run the code and see where is the problem. Could it be a typing issue ? (np.array vs list) ?
I usually achieve the same thing as you do with the following synthax (closer to `np.interp` synthax). Don't know if it can help:

```python
opti = cas.Opti()
n = 101
x1 = np.linspace(0, 2, n).tolist()
y1 = [non_casadi_function_1(i) for i in x1]

x2 = np.linspace(-1, 1, n).tolist()
y2 = [non_casadi_function_2(i) for i in x2]

m = 10
x1_opt = opti.variable(m)
x2_opt = opti.variable(m)
F1_interp= cas.interpolant('cs_interp','linear',[n],1,{"inline": True})
F1 = F1_interp(x1_opt,x1,y1)

F2_interp= cas.interpolant('cs_interp','linear',[n],1,{"inline": True})
F2 = F2_interp(x2_opt,x2,y2)
```

**Reply by mrrezaie** [reply]
Thanks for your response. This is actually a small part of my code and I don't know how to create a working example. But let me try your suggestions, I'll be back ...

**Reply by mrrezaie** [reply]
> Could it be a typing issue ? (np.array vs list) ?

all x1, x2, y1, and y2 are lists here.

> `F1 = F1_interp(x1_opt,x1,y1)`

It raises dimension mismatched error.

I also feel that some sort of memory leak is happening when running my code, because the size of my pagefile is being increased (~3.75 GB). I have 4 opti variables (73*80 matrix), 4 interpolants and lots of constraints:
![image](https://github.com/casadi/casadi/assets/38867713/2f737b00-ec51-4172-8b3e-84ac91ea46e5)
But this doesn't happen when I use my former code with mathematical functions (not `interpolant`s). 
Thanks again.

**Reply by sandeep026** [reply]
There is no objective in the code specified. A system of linear equations is solved.

**Reply by mrrezaie** [reply]
It won't be converged indeed. Based on your comments, it seems that interpolant functions with this form are differentiable by casadi and should work without issue. I will try to create a minimal working example with mathematical vs interpolant functions for comparison and share it here. Thanks for your help.

**Comment by mrrezaie** [comment]
Hi all, I created a working example with real data; please find it here: https://github.com/mrrezaie/testCASADI. I would be greatly grateful if you could take a look and let me know what is wrong with the interpolant functions. The formulation can be changed in the lines 5-6. Everything is well with mathematical functions. When running interpolants, pay attention to the memory usage. Thank you in advance.

**Reply by mrrezaie** [reply]
Hi, any suggestion on how to get the interpolants working? Have I implemented them correctly? Any help is greatly appreciated.

**Reply by mrrezaie** [reply]
> https://github.com/mrrezaie/testCASADI

I saw a few examples ([link](https://github.com/MarcoPenne/AMR_MPC-CBF/blob/33862ab3c3163967543dcf26eb0e8d825755456f/continuos_time/CarModel.py#L27), [link](https://github.com/YznMur/mown-project/blob/448dce8b1a66346cffb55d375c290b5f7bb4b23b/planning/move_controller/src/move_controller/move_nmpc.py#L176), [link](https://github.com/MIT-REALM/density_planner/blob/2c3b6d9b7bb6e62546a866ada151d2ec61021a18/motion_planning/MotionPlannerNLP.py#L402)) and I believe that I'm implementing the interpolants exactly the same as they did. The only difference I can see is that they used only one interpolant for constraints while I use [four](https://github.com/mrrezaie/testCASADI/blob/main/testCASADI.py#L141-L143) at the same time. Could it be the reason? Or is some data type conversion required?

My example can be executed on Google Colab as well if you wish not to download and run it locally:
```python
import os
!pip install casadi
!git clone https://github.com/mrrezaie/testCASADI
os.chdir('./testCASADI')
!python testCASADI.py
```
I would be greatly grateful for your help.

**Reply by mrrezaie** [reply]
Does the data in interpolant function matter? This is one of my curves the interpolant produces:
![image](https://github.com/casadi/casadi/assets/38867713/959d8d2e-f596-4eba-8094-da825563058f)
Perhaps this is not differentiable?! Any help is greatly appreciated.
